### PR TITLE
task(many): Enable eslint and fix linting errors

### PR DESCRIPTION
**Pull request by dschom**
## Because

- Not all workspaces were configured for linting
- Not all workspaces passed linting

## This pull request

- Makes sure the lint command runs all workspaces
- Applies standard eslint config across the board
- Fixes as many linting errors as possible
- Follows still needed for payments & settings

## Issue that this pull request solves

Closes: FXA-6793

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Other information (Optional)

Follows up have been filed for payments and settings due to a high number of eslint errors.

### CODE REVIEW

File: packages/fxa-dev-launcher/package.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "version": "1.0.0",
   "description": "",
   "scripts": {
+    "lint": "eslint . --ext .mjs",
```
Comment (pdehaan) [inline packages/fxa-dev-launcher/package.json:6]: :+1:
```js
Map(5) {
  '.eslintignore' => 1,
  '.json' => 2,
  '.md' => 1,
  '.mjs' => 2,
  '.nsprc' => 1
}
```

File: packages/fxa-auth-client/package.json [hunk 2]
```diff
@@ -13,6 +13,7 @@
     "./lib/": "./lib/"
   },
   "scripts": {
+    "lint": "eslint . --ext .ts",
```
Comment (pdehaan) [inline packages/fxa-auth-client/package.json:16]: :+1:
```js
Map(3) {
  '.eslintignore' => 1,
  '.json' => 4,
  '.ts' => 11
}
```

File: packages/fxa-auth-server/package.json [hunk 3]
```diff
@@ -16,7 +16,7 @@
     "bump-template-versions": "node scripts/template-version-bump",
     "audit": "npm audit --json | audit-filter --nsp-config=.nsprc --audit=-",
     "l10n-prime": "yarn l10n:prime fxa-auth-server",
-    "lint": "eslint . .storybook --ignore-pattern 'dist' --ext .js,.ts",
+    "lint": "eslint . .storybook --ext .js,.ts",
```
Comment (pdehaan) [inline packages/fxa-auth-server/package.json:19]: :+1:
```js
Map(21) {
  '.css' => 1,
  '.docker' => 1,
  '.eslintignore' => 1,
  '.eslintrc' => 2,
  '.ftl' => 73,
  '.graffle' => 1,
  '.handlebars' => 1,
  '.ini' => 3,
  '.js' => 365,
  '.json' => 143,
  '.lua' => 6,
  '.md' => 22,
  '.mjml' => 73,
  '.nsprc' => 1,
  '.prettierignore' => 1,
  '.py' => 1,
  '.scss' => 7,
  '.sh' => 4,
  '.sql' => 3,
  '.ts' => 185,
  '.txt' => 81
}
```

**Comment by pdehaan** [comment]
where my silly little file extension checker looks like this:

<details>
<summary>ext-grep.mjs</summary>

```js
import cp from 'node:child_process';
import path from 'node:path';
import { inspect } from 'node:util';

const pkgName = 'fxa-dev-launcher';

const files = gitLsFiles(pkgName);
console.log(inspect(extMap(files), { sorted: true }));

function gitLsFiles(name = '') {
  const files = cp
    .execSync(`git ls-files packages/${name}`)
    .toString()
    .trim()
    .split('\n');
  return files;
}

function extMap(files = []) {
  return files.reduce((map, filename) => {
    let ext = path.extname(filename);
    if (!ext) {
      ext = path.basename(filename);
    }
    const count = map.get(ext) || 0;
    if (ext.startsWith('.')) map.set(ext, count + 1);
    return map;
  }, new Map());
}
```
</details>

File: packages/functional-tests/.eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "extends": [
+    "react-app",
+    "react-app/jest"
+  ],
```
[the remark this answers is not in the document]
Comment (dschom) [reply packages/functional-tests/.eslintrc.json:5]: Ah yeah, I was just copying this around. That's a good call out. Linting does pass, but I'll remove the react extensions.

File: packages/fxa-graphql-api/src/auth/session-token.strategy.spec.ts [hunk 5]
```diff
@@ -4,17 +4,19 @@
...
 };
-let mockAuthClient = {
+const mockAuthClient = {
   deriveHawkCredentials: jest.fn(),
 };
+
 jest.mock('fxa-auth-client', () => mockAuthClient);
 jest.mock('fxa-shared/db/models/auth/session-token', () => mockSession);
 
+// eslint-disable-next-line import/first
```
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-graphql-api/src/auth/session-token.strategy.spec.ts:19]: Tried that. It breaks the mock.

File: packages/functional-tests/.eslintrc.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (dschom) [reply packages/functional-tests/.eslintrc.json]: Hmm, changing this broke white a few things. I am going to file a follow up for functional tests.

[the remark this answers is not in the document]
Comment (pdehaan) [reply packages/functional-tests/.eslintrc.json]: We can leave it as React for now, i guess having SOME ESLint rules is better than 0 rules.
I can also do a few quick greps or scripts to find all of our nested ESLint configs (.eslintrc, .eslintrc.js, .eslintrc.json, or buried in package.jsons) and see which of them are extending which groups of rules and how many are using the shared https://github.com/mozilla/fxa/tree/main/packages/eslint-plugin-fxa configs.

File: packages/fxa-customs-server/test/remote/too_many_bad_logins.js [hunk 6]
```diff
@@ -108,48 +108,48 @@ test('too many failed logins from the same email', async (t) => {
   const IP3 = '192.0.2.13';
 
   // eslint-disable-next-line no-unused-vars
```
Comment (pdehaan) [inline packages/fxa-customs-server/test/remote/too_many_bad_logins.js:110]: Q: Does this still need the `// eslint-disable-next-line no-unused-vars` now that we have a sparse array?

**Comment by pdehaan** [comment]
Interesting @pdehaan fact: Looks like the monorepo has 417 instances of `eslint-disable` (125), `eslint-disable-line` (150), and `eslint-disable-next-line` (142):

```sh
git grep -Eoh "eslint-disable((-next)?-line)?.*$" | sort | wc -l # 417
git grep -Eoh "eslint-disable((-next)?-line)?.*$" | sort | uniq | wc -l # 71

git grep -E "eslint-disable[^-]" | wc -l # 125
git grep -E "eslint-disable-line" | wc -l # 150
git grep -E "eslint-disable-next-line" | wc -l # 142
```

File: packages/fxa-event-broker/src/queueworker/service-notification.interface.ts [hunk 7]
```diff
@@ -51,6 +51,7 @@ function multiSchemaAttempt(
   return;
 }
 
+// eslint-disable-next-line @typescript-eslint/no-redeclare
 export const ServiceNotification = {
```
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-event-broker/src/queueworker/service-notification.interface.ts:55]: Yes, I was thinking we file a separate issue for this. And we should agree to follow this linting rule. I started trying to fix this, but it became a task in itself.

**Comment by dschom** [comment]
@pdehaan I think filing a separate task for addressing `eslint-disable*` is worth while. What do you think? Can you provide a break down per package? This might help us group these issues into reasonably sized tickets on package by package basis.

**Comment by pdehaan** [comment]
> Can you provide a break down per package?

/giphy "challenge accepted"
I'll try converting my bash greps into a `child_process.execSync()` and running through `lodash.groupby` and see if I can easily slice the 417 by package name. Shouldn't be too difficult, and then I'll file X tasks in Jira today. :+1:

**Comment by pdehaan** [comment]
@dschom Filed as FXA-7264 (which I think is randomly a spike, but I did link the 12 child tasks to that issue).
TLDR: I have no idea what I'm doing, but somebody with greater Jira chops than I can move/modify/close those as needed.

File: packages/fxa-auth-client/lib/client.ts [hunk 8]
```diff
@@ -151,6 +151,20 @@ function cleanStringify(value: any) {
   return JSON.stringify(value, (_, v) => (v == null ? undefined : v));
 }
 
+export class AuthClientError extends Error {
+  code: number;
+  errno: number;
+  error: string;
+
+  constructor(error: string, message: string, errno: number, code: number) {
```
[the remark this answers is not in the document]
Comment (chenba) [reply packages/fxa-auth-client/lib/client.ts:159]: If `result` has an `errno` then it's most likely an AppError from the auth-server.

File: packages/fxa-auth-client/lib/crypto.ts [hunk 9]
```diff
@@ -222,7 +222,7 @@ export async function jweDecrypt(
   jwe: string
 ): Promise<string> {
   const encoder = new TextEncoder();
-  const [header, empty, iv, ciphertext, authenticationTag] = jwe.split('.');
+  const [header, , iv, ciphertext, authenticationTag] = jwe.split('.');
```
Comment (chenba) [inline packages/fxa-auth-client/lib/crypto.ts:225]: Interesting that earlier I saw `undefined` being added but here we just leave it empty.  Linter need to make up its mind.

File: packages/fxa-auth-client/lib/hawk.ts [hunk 10]
```diff
@@ -46,7 +46,7 @@ export async function deriveHawkCredentials(token: hexstring, context: string) {
 
 function parseUri(input: string) {
   const parts = input.match(
-    /^([^:]+)\:\/\/(?:[^@/]*@)?([^\/:]+)(?:\:(\d+))?([^#]*)(?:#.*)?$/
+    /^([^:]+):\/\/(?:[^@/]*@)?([^/:]+)(?::(\d+))?([^#]*)(?:#.*)?$/
```
Comment (chenba) [inline packages/fxa-auth-client/lib/hawk.ts:49]: errr is this a linting fix?

File: packages/fxa-event-broker/src/jwtset/jwtset.service.spec.ts [hunk 11]
```diff
@@ -64,26 +64,28 @@ describe('JwtsetService', () => {
...
-        const payload = await PUBLIC_JWT.verify(token);
-        expect(payload.aud).toBe(TEST_CLIENT_ID);
-        expect(payload.sub).toBe('uid1234');
-        expect(payload.iss).toBe('test');
-        if (event === PROFILE_CHANGE_EVENT) {
-          expect(payload.events[PROFILE_EVENT_ID].email).toBe(
-            'test@mozilla.com'
-          );
-        }
+        await checkSet(event, method);
```
Comment (chenba) [inline packages/fxa-event-broker/src/jwtset/jwtset.service.spec.ts:88]: Was this a linting change?

**Review by chenba — APPROVED** [review]
Thanks for all the fixes, Dan.

File: packages/fxa-event-broker/src/queueworker/service-notification.interface.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (pdehaan) [reply packages/fxa-event-broker/src/queueworker/service-notification.interface.ts:55]: Ah, looks like we might only do this in 3 places currently:
```sh
git grep -nA1 " @typescript-eslint/no-redeclare" packages | cat

packages/fxa-auth-server/lib/types.ts:87:// eslint-disable-next-line @typescript-eslint/no-redeclare
packages/fxa-auth-server/lib/types.ts-88-export const AuthLogger = new Token<AuthLogger>('AUTH_LOGGER');
--
packages/fxa-auth-server/lib/types.ts:91:// eslint-disable-next-line @typescript-eslint/no-redeclare
packages/fxa-auth-server/lib/types.ts-92-export const ProfileClient = new Token<ProfileClient>('PROFILE_CLIENT');
--
packages/fxa-event-broker/src/queueworker/service-notification.interface.ts:54:// eslint-disable-next-line @typescript-eslint/no-redeclare
packages/fxa-event-broker/src/queueworker/service-notification.interface.ts-55-export const ServiceNotification = {
```

File: packages/fxa-auth-client/lib/client.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-auth-client/lib/client.ts:159]: @pdehaan Thanks for double checking this. I did my best to ensure that the mapping was the same. So 
```
result -> message
error -> error
errno -> errorno
response.status -> code
```

See L209 in the diff. I put the constructor args in the same order as the object to ensure something didn't shift.

File: packages/fxa-auth-client/lib/hawk.ts (hunk 10 above)
Comment (dschom) [reply packages/fxa-auth-client/lib/hawk.ts:49]: Yep this was useless escape.

Comment (dschom) [reply packages/fxa-auth-client/lib/hawk.ts:49]: @chenba Do you have any concerns with this?

Comment (chenba) [reply packages/fxa-auth-client/lib/hawk.ts:49]: I do not.  👍

File: packages/fxa-event-broker/src/jwtset/jwtset.service.spec.ts (hunk 11 above)
Comment (dschom) [reply packages/fxa-event-broker/src/jwtset/jwtset.service.spec.ts:88]: Yep, it was barking about, `Function declared in a loop contains unsafe references to variable(s) 'service'.eslintno-loop-func`

File: packages/fxa-event-broker/src/queueworker/service-notification.interface.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-event-broker/src/queueworker/service-notification.interface.ts:55]: Yeah. Once we fix those, we then run into in the issue of what to name the re declarations, and then run into the issue updating references to these renames. I've filed a follow up. https://mozilla-hub.atlassian.net/browse/FXA-7267

File: packages/fxa-auth-client/lib/crypto.ts (hunk 9 above)
Comment (pdehaan) [reply packages/fxa-auth-client/lib/crypto.ts:225]: > _The next step will be to start arguing about what rules to enforce 😛 and applying them consistently._

"semicolon has entered the chat"
